### PR TITLE
Non-scalar Multinomial draws

### PR DESCRIPTION
**Pull request by jinze1994**
This PR tries to fix the issue in [#12804](https://github.com/tensorflow/tensorflow/issues/12804) where supports total_count to be a non-scalar tensor and broadcasts by `map_fn`.

The reason why I use `map_fn` to broadcast manually is that current `tf.multinomial` sampler op don't support to produce different total_count within a batch, and `map_fn` can enhance this underlying op effectively without modifying it.

It's a little dilemma if we modify the underlying op to support different total_count within a batch, because that will produce variable-length draws within a batch, unless we add a new function with different interface for multinomial sampler.

### CODE REVIEW

**Comment by jinze1994** [comment]
@jvdillon Could you please review my code? Thank you very much~

**Comment by jvdillon** [comment]
Looking now...

File: tensorflow/python/ops/distributions/multinomial.py [hunk 1]
```diff
@@ -230,29 +233,35 @@ def _event_shape(self):
...
-        seed=seed)
-    x = array_ops.reshape(x, shape=[-1, n, n_draws])
-    x = math_ops.reduce_sum(array_ops.one_hot(x, depth=k),
-                            axis=-2)  # shape: [B, n, k]
+
+    # boardcast the total_count and logits to same shape
+    n_draws = array_ops.ones_like(
+        self.logits[..., 0], dtype=n_draws.dtype) * n_draws
+    logits = array_ops.ones_like(
+        n_draws[..., None], dtype=self.logits.dtype) * self.logits
```
Comment (jvdillon) [inline tensorflow/python/ops/distributions/multinomial.py]: Please use array_ops.newaxis rather than `None`

File: tensorflow/python/ops/distributions/multinomial.py [hunk 2]
```diff
@@ -230,29 +233,35 @@ def _event_shape(self):
...
+
+    # flatten the total_count and logits
+    flat_logits = array_ops.reshape(logits, [-1, k]) # [B1*B2*...*Bm, k]
```
Comment (jvdillon) [inline tensorflow/python/ops/distributions/multinomial.py]: Please use kwarg for literals, ie, 
flat_logits = array_ops.reshape(logits, shape=[-1, k]) # [B1*B2*...*Bm, k]

**Comment by jvdillon** [comment]
Overall, looks good--just a few nitpicks.   I had actually hoped we could change the underlying C++ sampler, but I suppose this is a reasonable stop-gap measure.  I'll approve once the minor style issues have been addressed.

**Review by jvdillon — CHANGES_REQUESTED** [review]

**Comment by jinze1994** [comment]
@jvdillon 

Thank you for getting this reviewed, I have modified the code according to your suggestion. By the way, I am glad to try to help changing the underlying sampler in the future.

**Comment by sb2nov** [comment]
@jvdillon is this good to go?

**Review by jvdillon — APPROVED** [review]

**Comment by jvdillon** [comment]
Sorry for my slowness to review.  Thank you for your contribution.  Please feel free to reach out to me regarding issues that come up in your improvement in the underlying multinomial sampler.  It can also be made much more efficient than it currently is.

**Comment by sb2nov** [comment]
Jenkins, test this please.

**Comment by sb2nov** [comment]
@jinze1994 there seems to be test failure with the change https://ci.tensorflow.org/job/tensorflow-pull-requests-cpu-python3/6702/consoleFull

**Comment by jinze1994** [comment]
@jvdillon @sb2nov 

Sorry for my delay to handle the test failure. I think this test failure occurred because there were not enough sample time when check the covariance from sampling, which led to some deviation. And I fix it by augmenting the sample time.

As for improving in the underlying multinomial sampler, thanks for jvdillon's support first, but I am not sure if I have enough time to achieve that. If I have enough time in the future, I will create a new PR for modify the underlying sampler.

**Comment by jvdillon** [comment]
Thanks again!  (Right, fixing the C++ sampler should definitely be a separate PR, when/if you get to it.)

**Comment by jinze1994** [comment]
@sb2nov @jvdillon 
My pleasure. So could you please test this PR again?

**Comment by jinze1994** [comment]
@jvdillon Hi bro~ Could you please test this PR again? Or is there anything I could do for this PR?

**Comment by sb2nov** [comment]
Jenkins, test this please.

**Comment by jinze1994** [comment]
I noticed that there ware still some test failure, and I think that occurred because there were not enough sample times, because I passed the test locally most of time.

I have already augmented the sample times at the failure place. Please test it again? Sorry for having troubled you so much. Thank you very much. @sb2nov @jvdillon

**Comment by jinze1994** [comment]
@jvdillon @sb2nov 
I've already handled the test failure, could you please test it again? Thank you very much.

**Comment by jvdillon** [comment]
Jenkins, test this please.

**Comment by jvdillon** [comment]
How are things coming along? Anything I can help with?

**Review by jvdillon — COMMENTED** [review]
Assuming tests pass, this LGTM.

**Comment by vrv** [comment]
@tensorflow-jenkins test this please

**Comment by jinze1994** [comment]
@jvdillon @vrv 
I haven't figured out the failure reason of those check above, but it seems not to be related with this PR.
So is there anything I can do to avoid it and modify?

**Comment by jvdillon** [comment]
Breakages seem unrelated.

**Comment by jinze1994** [comment]
@jvdillon @vrv Is this good to merge?

**Comment by jvdillon** [comment]
@vrv seems like the breakages are unrelated. Maybe kick off another run and see if they all pass this time?

**Comment by vrv** [comment]
@tensorflow-jenkins test this please

**Comment by jinze1994** [comment]
@jvdillon 
Sorry for that there are still some test failure, but I don't know how to handle that.
Should I merge with the current master branch to repair those BUILD failure?

**Comment by gunan** [comment]
Jenkins, test this please.

**Comment by gunan** [comment]
Jenkins, test this please.

**Comment by jinze1994** [comment]
@jvdillon @vrv 
I have rebased this PR with current tensorflow master, and it should solve the BUILD failures. So please test it again. Thank you.

**Comment by gunan** [comment]
Jenkins, test this please.

**Comment by jinze1994** [comment]
Those failures seem unrelated with this PR.
